File: go/core/fs/fs.go
package fs

import (
	"fmt"
	"os"
	"io"
	"strings"
	"path/filepath"
)

func PrepareDirectories(distDir string) error {
	if err := os.RemoveAll(distDir); err != nil {
		return fmt.Errorf("failed to clear destination directory %s: %w", distDir, err)
	}
	if err := os.MkdirAll(distDir, os.ModePerm); err != nil {
		return fmt.Errorf("failed to create destination directory %s: %w", distDir, err)
	}
	return nil
}

func CollectMarkdownFiles(srcDir string) ([]string, error) {
	var files []string
	err := filepath.Walk(srcDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if strings.HasSuffix(path, ".md") {
			files = append(files, path)
		}
		return nil
	})
	return files, err
}

func CopyAssets(src, dest string) error {
	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		destPath := filepath.Join(dest, strings.TrimPrefix(path, src))
		if info.IsDir() {
			return os.MkdirAll(destPath, os.ModePerm)
		}
		return copyFile(path, destPath)
	})
}

func copyFile(srcPath, destPath string) error {
	srcFile, err := os.Open(srcPath)
	if err != nil {
		return fmt.Errorf("failed to open source file %s: %w", srcPath, err)
	}
	defer srcFile.Close()
	destFile, err := os.Create(destPath)
	if err != nil {
		return fmt.Errorf("failed to create destination file %s: %w", destPath, err)
	}
	defer destFile.Close()
	_, err = io.Copy(destFile, srcFile)
	return err
}

File: src/fs.rs
use anyhow::{Context, Result};
use std::fs;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

pub fn prepare_directories(dist_dir: &Path) -> Result<()> {
    if dist_dir.exists() {
        fs::remove_dir_all(dist_dir).context("failed to clear destination directory")?;
    }
    fs::create_dir_all(dist_dir).context("failed to create destination directory")?;
    Ok(())
}

pub fn collect_djot_files(src_dir: &Path) -> Result<Vec<PathBuf>> {
    let mut files = Vec::new();
    for entry in WalkDir::new(src_dir).into_iter().filter_map(|e| e.ok()) {
        let path = entry.path();
        if path.is_file() && path.extension().is_some_and(|ext| ext == "djot") {
            files.push(path.to_path_buf());
        }
    }
    Ok(files)
}

pub fn copy_assets(src: &Path, dest: &Path) -> Result<()> {
    if !src.exists() {
        return Ok(());
    }

    for entry in WalkDir::new(src).into_iter().filter_map(|e| e.ok()) {
        let path = entry.path();
        let relative_path = path.strip_prefix(src).context("failed to strip prefix")?;
        let dest_path = dest.join(relative_path);

        if path.is_dir() {
            fs::create_dir_all(&dest_path).context("failed to create asset directory")?;
        } else if path.is_file() {
            fs::copy(path, &dest_path).context("failed to copy asset file")?;
        }
    }
    Ok(())
}
